layer/layer_store.go

```diff
@@ -657,3 +657,121 @@ func (n *naiveDiffPathDriver) DiffGetter(id string) (graphdriver.FileGetCloser,
 	}
 	return &fileGetPutter{storage.NewPathFileGetter(p), n.Driver, id}, nil
 }
+
+func (ls *layerStore) CopyAndMerge(id, parent ChainID) (Layer, error) {
+	l, err := ls.Get(id)
+	if err != nil {
+		return nil, err
+	}
+	defer ls.Release(l)
+	var p *roLayer
+	var pid string
+	if len(parent) > 0 {
+		p = ls.get(parent)
+		if p == nil {
+			return nil, ErrLayerDoesNotExist
+		}
+		pid = p.cacheID
+	}
+
+	if len(parent) > 0 {
+		var isParent bool
+		ll := l
+		for {
+			p := ll.Parent()
+			if p == nil {
+				break
+			}
+			if p.ChainID() == parent {
+				isParent = true
+				break
+			}
+			ll = p
+		}
+		if !isParent {
+			return nil, fmt.Errorf("not a parent")
+		}
+	}
+
+	var tarStreams []io.ReadCloser
+	defer func() {
+		for _, ts := range tarStreams {
+			ts.Close()
+		}
+	}()
+
+	for {
+		ts, err := l.TarStream()
+		if err != nil {
+			return nil, err
+		}
+
+		tarStreams = append([]io.ReadCloser{ts}, tarStreams...)
+
+		l = l.Parent()
+		if l == nil || l.ChainID() == parent {
+			break
+		}
+	}
+
+	// Create new roLayer
+	layer := &roLayer{
+		parent:         p,
+		cacheID:        stringid.GenerateRandomID(),
+		referenceCount: 1,
+		layerStore:     ls,
+		references:     map[Layer]struct{}{},
+	}
+
+	if err = ls.driver.Create(layer.cacheID, pid, "", nil); err != nil {
+		return nil, err
+	}
+
+	tx, err := ls.store.StartTransaction()
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err != nil {
+			logrus.Debugf("Cleaning up layer %s: %v", layer.cacheID, err)
+			if err := ls.driver.Remove(layer.cacheID); err != nil {
+				logrus.Errorf("Error cleaning up cache layer %s: %v", layer.cacheID, err)
+			}
+			if err := tx.Cancel(); err != nil {
+				logrus.Errorf("Error canceling metadata transaction %q: %s", tx.String(), err)
+			}
+		}
+	}()
+
+	for _, ts := range tarStreams {
+		if err = ls.applyTar(tx, ts, string(parent), layer); err != nil {
```

Comment:
This function will not work in aufs and overlay2 since they do a direct unpack of the tar into the diff directory. The expectation with these drivers is that the `applyTar` is called on an empty layer and the contents can be unpacked into that layer without the worry of overwriting existing content. This is especially important related to whiteouts which are written to disk for these drivers rather than having the action of the whiteout applied. A possible work around would be to simulate what the naive diff driver is doing here by creating the new layer, mounting it, then calling the ApplyLayer functionality. However I am not a fan of doing any layer content manipulation in the layer store, just only thing that immediately comes to mind.

AUFS and Overlay2 use https://github.com/docker/docker/blob/master/pkg/archive/archive.go#L700

NaiveDiffDriver uses https://github.com/docker/docker/blob/master/pkg/archive/diff.go#L22